tests/Integration/Behaviour/Features/Context/Domain/Discount/AddDiscountFeatureContext.php

```diff
@@ -0,0 +1,85 @@
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace Tests\Integration\Behaviour\Features\Context\Domain\Discount;
+
+use Behat\Gherkin\Node\TableNode;
+use PrestaShop\PrestaShop\Core\Domain\Discount\Command\AddFreeShippingDiscountCommand;
+use PrestaShop\PrestaShop\Core\Domain\Discount\Exception\DiscountConstraintException;
+use PrestaShop\PrestaShop\Core\Domain\Discount\Exception\DiscountNotFoundException;
+
+class AddDiscountFeatureContext extends AbstractDiscountFeatureContext
+{
+    /**
+     * @When I create a free shipping discount :discountReference with following properties:
+     * @When there is a free shipping discount :discountReference with following properties:
+     *
+     * @param string $discountReference
+     * @param TableNode $node
+     */
+    public function createFreeShippingDiscountIfNotExists(string $discountReference, TableNode $node): void
+    {
+        $data = $this->localizeByRows($node);
+        $this->checkIfDiscountExist($discountReference, $data);
+        try {
+            $command = new AddFreeShippingDiscountCommand($data['name']);
+            $this->createDiscountIfNotExist($discountReference, $data, $command);
+        } catch (DiscountConstraintException $e) {
+            $this->setLastException($e);
+        }
+    }
+
+    /**
+     * @Then discount :discountReference should have the following properties:
+     *
+     * @param string $cartRuleReference
+     * @param TableNode $tableNode
+     */
+    public function assertDiscount(string $cartRuleReference, TableNode $tableNode): void
+    {
+        $this->assertDiscountProperties(
+            $this->getDiscountForEditing($cartRuleReference),
+            $this->localizeByRows($tableNode)
+        );
+    }
+
+    private function checkIfDiscountExist(string $discountReference, array $data): void
+    {
+        if ($this->getSharedStorage()->exists($discountReference)) {
+            try {
+                // if discount already exists we assert all its expected properties
+                $this->assertDiscountProperties(
+                    $this->getDiscountForEditing($discountReference),
+                    $data
+                );
+
+                return;
+            } catch (DiscountNotFoundException $e) {
+                // if discount was not found we proceed to create it under this reference.
+            }
+        }
+    }
```

Comment:
I think you can remove this part, the Behat context doesn't need to check this It simply executes the action written in the scenario so it's the scenario's author that needs to write a scenario that makes sense with existing entities